### PR TITLE
Restore tf_prefix support for ROS2

### PR DESCRIPTION
**Pull request by smnogar**
Relevant issue: https://github.com/ros/robot_state_publisher/issues/147
Relevant PRs: Noetic support: https://github.com/ros/robot_state_publisher/pull/139 Closed ROS2 PR: https://github.com/ros/robot_state_publisher/pull/127

This is a quick and dirty implementation of tf_prefix for foxy. Once the string formatting issues are resolved in https://github.com/ros/robot_state_publisher/pull/139, I recommend adding them here. If this is too duplicative feel free to close this.

### CODE REVIEW

**Comment by clalancette** [comment]
@smnogar First of all, I apologize for the very long delay in looking at this.

What I ended up doing here was make some changes to your PR, partially based off of the discussion that has been going on in #139.  In particular I did the following:

1. I rebased this on top of the `ros2` branch, which is where we land any new changes first.  Once we have it there, we can consider backporting it to Galactic and Foxy.
1. I renamed the prefix from `tf_prefix` to `frame_prefix`.  See https://github.com/ros/robot_state_publisher/pull/139#pullrequestreview-506804970 for why we think this should not just be `tf_prefix`.
1. I removed any special handling of `/` here.  Since this is now just a pure prefix, and the slash doesn't mean anything special to tf2 anymore, there is no reason to handle it specially here.
1. I added in a test for this to ensure we are getting what we expect.

With that, I'm pretty happy with this.  I'd appreciate a look from you to make sure this meets your needs.  @sloretz When you have some time, a review here would be appreciated as well.

File: test/test_two_links_fixed_joint_prefix.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include <gtest/gtest.h>
+
+#include <rclcpp/rclcpp.hpp>
+#include <sensor_msgs/msg/joint_state.hpp>
```
Comment (sloretz) [inline test/test_two_links_fixed_joint_prefix.cpp]: nit, unused include?

File: CMakeLists.txt [hunk 2]
```diff
@@ -80,6 +80,16 @@ if(BUILD_TESTING)
   add_launch_test(test/two_links_fixed_joint-launch.py
     ARGS "test_exe:=$<TARGET_FILE:test_two_links_fixed_joint>")
 
+  add_executable(test_two_links_fixed_joint_prefix test/test_two_links_fixed_joint_prefix.cpp)
+  ament_target_dependencies(test_two_links_fixed_joint_prefix
+    rclcpp
+    sensor_msgs
```
Comment (sloretz) [inline CMakeLists.txt]: nit, unnecessary dependency?

**Review by sloretz — APPROVED** [review]
A couple nitpicks, otherwise LGTM!

File: test/test_two_links_fixed_joint_prefix.cpp (hunk 1 above)
Comment (clalancette) [reply test/test_two_links_fixed_joint_prefix.cpp]: Yes, definitely.  Removed it from here and from `test_two_links_fixed_joint.cpp` in e83f247

File: CMakeLists.txt (hunk 2 above)
Comment (clalancette) [reply CMakeLists.txt]: Yes, definitely.  Removed it from here and from `test_two_links_fixed_joint` in e83f247

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14684/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9436/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12349/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14827/)

**Comment by clalancette** [comment]
All right, nits are fixed, CI is happy.  I'm going to go ahead and merge this one in.  @smnogar If you'd like to have this in Foxy or Galactic, please pipe up here and/or feel free to open backport PRs for those.  Thanks for the contribution.

**Comment by padhupradheep** [comment]
Is it possible to have the frame_prefix support for foxy as well ? I guess this is a cool feature, that we would love to have!

**Comment by clalancette** [comment]
> Is it possible to have the frame_prefix support for foxy as well ? I guess this is a cool feature, that we would love to have!

So looking back over this code, I realize that the big problem with backporting is that it is not ABI compatible.  That is, it changes the size of to `RobotStatePublisher` class, so we can't backport it as-is.

That said, we can probably do something slightly hacky where we don't declare the parameter, but instead allow undeclared parameters to be set and just get the parameter as needed.  That won't change the ABI.  Let me give that a shot and see if I can make it work.

**Pull request merged**